client/src/components/Home.js

```diff
@@ -65,60 +64,93 @@ const Home = ({ user, logout }) => {
+      let newState = [...conversations];
+      let currentConv = null
+      conversations.forEach((convo, index) => {
         if (convo.otherUser.id === recipientId) {
-          convo.messages.push(message);
+          convo.messages.unshift(message);
           convo.latestMessageText = message.text;
-          convo.id = message.conversationId;
+          currentConv = convo;
+          const idRandom = (Math.random() + 1).toString(36).substring(7);
+          currentConv.id = idRandom;
+          newState.splice(index, 1);
+          return;
         }
       });
-      setConversations(conversations);
+      console.log(currentConv);
+      if (currentConv != null) newState.unshift(currentConv);
+      setConversations(newState);
     },
     [setConversations, conversations]
   );
 
   const addMessageToConversation = useCallback(
-    (data) => {
-      // if sender isn't null, that means the message needs to be put in a brand new convo
-      const { message, sender = null } = data;
-      if (sender !== null) {
-        const newConvo = {
-          id: message.conversationId,
-          otherUser: sender,
-          messages: [message],
-        };
-        newConvo.latestMessageText = message.text;
-        setConversations((prev) => [newConvo, ...prev]);
-      }
+    (data, recipientId) => {
+      const { message, sender } = data;
+      const newState = [...conversations];
+      let currentConv = null;
 
-      conversations.forEach((convo) => {
-        if (convo.id === message.conversationId) {
-          convo.messages.push(message);
-          convo.latestMessageText = message.text;
+      if (recipientId === undefined) {
+        //recipient  
+        if (sender !== null) {
+          //new conversation
+          currentConv = {
+            id: message.conversationId,
+            otherUser: sender,
+            messages: [message],
+            latestMessageText: message.text,
+            user1: null,
+          };
+        } else {
+          newState.forEach((convo, index) => {
+            if (convo.id === message.conversationId && convo.otherUser.id === message.senderId) {
+                convo.messages.unshift(message);
+                convo.latestMessageText = message.text;
+                currentConv = convo;
+                newState.splice(index, 1);
+                return;
+              }
+          });
         }
-      });
-      setConversations(conversations);
+      } else {
+        //sender side
+        newState.forEach((convo, index) => {
+          if (convo.otherUser.id === recipientId) {
+            convo.messages.unshift(message);
+            convo.latestMessageText = message.text;
+            //remove this conv and set it as the first
+            currentConv = convo;
+            newState.splice(index, 1);
+            return;
+          }
+        });
+      }
+
+      if (currentConv != null) newState.unshift(currentConv);
+      setConversations(newState);
```

Comment:
Most of the comments I mentioned above apply here too (push vs unshift, changing the order of the conversations, and above all: not using a functional update).